core/src/main/scala/za/co/absa/spline/harvester/dispatcher/KafkaLineageDispatcher.scala

```diff
@@ -17,85 +17,63 @@
 package za.co.absa.spline.harvester.dispatcher
 
 import org.apache.commons.configuration.{Configuration, ConfigurationConverter}
-import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord}
-import org.apache.kafka.common.header.Header
+import org.apache.kafka.clients.producer.KafkaProducer
 import org.apache.spark.internal.Logging
 import za.co.absa.commons.config.ConfigurationImplicits.ConfigurationRequiredWrapper
-import za.co.absa.spline.harvester.dispatcher.httpdispatcher.ProducerApiVersion
+import za.co.absa.commons.version.Version
+import za.co.absa.spline.harvester.dispatcher.KafkaLineageDispatcher._
 import za.co.absa.spline.harvester.dispatcher.kafkadispatcher._
-import za.co.absa.spline.harvester.json.HarvesterJsonSerDe.impl._
+import za.co.absa.spline.harvester.dispatcher.modelmapper.ModelMapper
 import za.co.absa.spline.producer.model.v1_1.{ExecutionEvent, ExecutionPlan}
-import KafkaLineageDispatcher._
-
-import java.util.Properties
-import scala.collection.JavaConverters.asJavaIterableConverter
-import scala.util.control.NonFatal
 
 /**
  * KafkaLineageDispatcher is responsible for sending the lineage data to spline gateway through kafka
  */
-class KafkaLineageDispatcher(topic: String, producerProperties: Properties)
-  extends LineageDispatcher
-    with Logging {
+class KafkaLineageDispatcher(
+  apiVersion: Version,
+  createSenderFactory: Version => SplineRecordSenderFactory
+) extends LineageDispatcher
+  with Logging {
 
+  import za.co.absa.spline.harvester.json.HarvesterJsonSerDe.impl._
 
   def this(configuration: Configuration) = this(
-    configuration.getRequiredString(TopicKey),
-    ConfigurationConverter.getProperties(configuration.subset(ProducerKey))
+    Version.asSimple(configuration.getRequiredString(ApiVersion)),
+    apiVersion => new SplineRecordSenderFactory(
+      apiVersion,
+      configuration.getRequiredString(TopicKey),
+      () => new KafkaProducer(ConfigurationConverter.getProperties(configuration.subset(ProducerKey)))
+    )
   )
 
-  private val producer = new KafkaProducer[String, String](producerProperties)
-
-  logInfo(s"Kafka topic: $topic")
-
-  sys.addShutdownHook(producer.close())
-
   override def name = "Kafka"
 
-  private val planKafkaHeaders = Seq[Header](
-    new KafkaHeader(SplineKafkaHeaders.ApiVersion, ProducerApiVersion.Default.asString),
-    new KafkaHeader(SplineKafkaHeaders.TypeId, "ExecutionPlan")
-  )
+  private val senderFactory = {
+    val sf = createSenderFactory(apiVersion)
+    sys.addShutdownHook(sf.close())
```

Comment:
This seems weird. It looks like closing the factory. 

Also, the `AutoCloseable` there: I don't think a factory should really be handling closing of resources.

Reply:
it actually _is_ closing a factory. The rational here is AMD cycle management - who created a resource is also responsible to dispose it properly. Hence the factory also has it's own lifecycle and here we ask the factory to shutdown.

Reply:
> who created a resource is also responsible to dispose it properly

The factory is creating a resource, but KafkaLineageDispatcher has to dispose it. Why not call `sys.addShutdownHook` inside the factory?

What do you mean by AMD?

Reply:
> What do you mean by AMD?

Add, Modify, Delete

> Why not call `sys.addShutdownHook` inside the factory?

I didn't want a resource to outlive the factory. The idea here is to let the factory to manage the life cycle of the resource(s) it produces. So that from the user perspective, I call a factory to obtain a resource instance and just use it without having to think about if the returned resource is shared or isolated. All I need is to close a factory when I'm done and the factory will decide how to dispose its products properly.
Strictly speaking, I don't like `sys.addShutdownHook` here either as it basically leads to the same issue as if it were in the factory - the resource's and the factory's lifetime is longer that the dispatcher's one. Ideally I would want the dispatcher to have it's own life cycle management as well, but currently I don't see a good way of doing it.... apart maybe hooking the `SparkListener.onApplicationEnd()` event... The app still outlives the dispatcher and the session, but its lifecycle is shorter than the JVM one, thus it looks more appropriate hook than the JVM shutdown hook. What do you think? Since we have AgentBOM we can easily tie all agent components (dispatchers, filters etc) life cycle to the Spark context life cycle, and with that basically tie the entire Spline agent life cycle to it as well. So just in case several Spark contexts are started and stopped in one JVM the associated Spline agents would not leak any orphan resources.

Reply:
I think you are talking about other concern, `sys.addShutdownHook` is ok for this simple case I think.

My main point was that something named **factory** should be responsible solely for object creation and not it's destruction. Otherwise, it would be named **manager** or something else. Having the close method on the factory is making the code confusing.

Is the whole class really needed? It looks like a method in Dispatcher could do the same and would avoid this confusion completely.